Remove airloop flow reset which can cause max iteration errors and fix AirTerminal:SingleDuct:VAV schedule logic

mjwitte opened the pull request:
Addresses #4285 Allow AirTerminal:SingleDuct:VAV:Reheat to be scheduled off (schedule was not being checked).

Also, remove flow reset in air loop iteration loop after first HVAC iteration which could cause max iteration errors by resetting terminal unit flow rates to max flow limit regardless of schedules. This was a broader problem for certain configurations.

Notes for review:
Defect file 7781Test.idf should show no change before/after (design days only, run with no epw)

Defect file BslSMC20130916-NoAnnual-MinimalOutput.idf (also no epw) 
Should look like this for the summer day before fix:
![image](https://cloud.githubusercontent.com/assets/4991075/13680459/3bb5488a-e6be-11e5-98cc-f805b76b6480.png)

And like this after fix:
![image](https://cloud.githubusercontent.com/assets/4991075/13680391/fcd2da06-e6bd-11e5-9a9a-87ce0a9bb54f.png)

mjwitte committed:
Initial attempt to fix VAV terminal unit schedule problem

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 87382760-VAVTerminalUnitSchedule-Issue4285

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 87382760-VAVTerminalUnitSchedule-Issue4285

mjwitte committed:
Remove ResetTerminalUnitFlowLimits from HVACManager IterAir loop

mjwitte commented:
_21 Jan 2016 - See further down for latest results - this comment is superceded._

This causes big diffs in 7 files. (Updated on 27 Oct, some improvement with revised fix)

These need further work:
- [ ] 5ZoneWLHPPlantLoopTower - _Adds_ max iteration errors, <1% max relative diff _Oct 22, 2015 - new max iteration errors still there_
- [x] 5ZoneWarmestVFD_FCMAuto - ~~3% change in HVAC fan electricity~~ _Oct 22, 2015 - <1% change in all energy use for annual_
- [x] HeatPumpCycFanWithEcono ~~_adds_ max iteration errors~~ _Oct 27, 2015 - new max iteration errors are gone now_
- [ ] VSHeatPumpWaterToAirWithRHControl - _eliminates_ max iteration warnings, but _adds_ warmup convergence errors, <1 hr diffs in setpoint not met, small diffs in elec and gas - _Oct 22, 2015 - warmup convergence errors are gone with runperiod, still there for design days_
- [x] VSHeatPumpWaterToAirEquationFit - _eliminates_ max iteration warnings, ~~1% change in annual electricity, 2.5% change in heating electric~~ _Oct 22, 2015 - change in annual electricity and heating electricity now <1%_

These look OK:
ChangeoverBypassVAV - tiny relative diff in gas use
HeatPumpCycFanWithEcono - <1 hr change in setpoint not met/time not comfortable, small relative diffs in fan electric
UnitarySystem_VSHeatPumpWaterToAirEquationFit - _eliminates_ max iteration warnings, small relative diffs in fan electric
And 22 files had small diffs.

mjwitte commented:
5ZoneWLHPPlantLoopTower is an interesting example file (and so are all of the example files with AirLoopHVAC:UnitaryHeatPump:WaterToAir.  Most parts of the heat pump are hard-sized to a flow rate of 2.0, but the terminal units are autosized.  In this file and (most, all?) of the others, the heat pump is an air loop serving a single zone.  When run after the fix with cycling fan (as-is) it generates max iteration warnings (didn't before).  With continuous fan, no warnings.  With the terminal units sized to 2.0, no change (iteration warnings).  With everything autosized, it hangs (before and after this fix).

mjwitte commented:
5ZoneWarmestVFD_FCMAuto has duct leakage in the ADUs.  Setting this to zero makes the diffs much smaller.

mjwitte commented:
VSHeatPumpWaterToAirWithRHControl has humidity control, turning that off eliminates the warmup error.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 87382760-VAVTerminalUnitSchedule-Issue4285

Myoldmopar commented:
> These need further work

You made comments following this quote; did those comments wrap up your "further work" or investigation?

mjwitte commented:
@Myoldmopar As much as I would like to see this go in, none of the questionable changes listed at the top are completely resolved.  Probably should take the time to further understand and/or find a fix for those rather than slam this in.  Milestone changed.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 87382760-VAVTerminalUnitSchedule-Issue4285

mjwitte committed:
Only remove ResetTerminalUnitFlowLimits in HVACManager IterAir loop

mjwitte committed:
Resolve max iteration problem with cycling fan heat pump

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 87382760-VAVTerminalUnitSchedule-Issue4285

nrel-bot-3 commented:
@mjwitte @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@mjwitte @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@mjwitte @lgentile it has been 14 days since this pull request was last updated.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 87382760-VAVTerminalUnitSchedule-Issue4285

mjwitte committed:
Merge branch '87382760-VAVTerminalUnitSchedule-Issue4285' of https://github.com/NREL/EnergyPlus into 87382760-VAVTerminalUnitSchedule-Issue4285

mjwitte commented:
Current diff status.
Of the 5 files with ERR diffs, 4 of them eliminate max iteration warnings.
- [x] 5ZoneWLHPPlantLoopTower still adds a max iteration warning. _Fixed Jan 29_
- [ ] VSHeatPumpWaterToAirWithRHControl still eliminates max iteration warnings, but _adds warmup convergence errors_ for design days _New issue #5447 posted for this_

Of the 8 files with big ESO diffs, all of the diffs are small (<1% or tiny quantities to begin with) except
- [ ] VSHeatPumpWaterToAirWithRHControl shows a large diff in cooling coil load _New issue #5447 posted for this_

Will run annual diffs of the 8 files.

mjwitte commented:
Annual diffs:
5ZoneWarmestVFD_FCMAuto:  <1% annual energy, a few hours change in tstat hrs unmet
5ZoneWaterLoopHeatPump: small diffs, ~~but adds new max iteration and oscillating node mass flow rate warnings~~ _Fixed Jan 29_
5ZoneWLHPPlantLoopTower:  tiny annual diffs,  ~~but adds new max iteration and oscillating node mass flow rate warnings~~ _Fixed Jan 29_
UnitarySystem_5ZoneWaterLoopHeatPump: small diffs

_The next four all have max iteration warnings with current develop, so new issue #5447  posted to deal with them_
UnitarySystem_VSHeatPumpWaterToAirEquationFit: 3% diff in peak cooling electricity, everything else is small diffs,  fewer max iterations
VSHeatPumpWaterToAirEquationFit: ditto,
VSHeatPumpWaterToAirWithRHControl: <1% change in annual energy, a few hours difference in setpoint not met,  fewer max iterations (80 vs 989) _plus warmup convergence error_
HVACTemplate-5ZoneUnitarySystem:  25% increase in heating electricity, big drop in cooling hrs not met (from 232 to 8), _but still has max iteration warnings and more reports of mismatches_

mjwitte committed:
Cleanup water loop heat pump example files

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 87382760-VAVTerminalUnitSchedule-Issue4285

mjwitte commented:
Stopping here - just need a unit test now.

nrel-bot-3 commented:
@mjwitte @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-2 commented:
@mjwitte @lgentile it has been 14 days since this pull request was last updated.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 87382760-VAVTerminalUnitSchedule-Issue4285

mjwitte committed:
VAV terminal unit add schedule check for cooling mode plus unit tests

mjwitte committed:
Remove IDD change (save for post 8.5 release)

mjwitte commented:
@Myoldmopar This should be complete - will look at CI results - than tag you and/or @rraustad for review.  Note this overlaps with #5532, so test the "before" without either one of these.

mjwitte committed:
Fix build warning in SingleDuct.unit.cc

mjwitte commented:
@Myoldmopar @rraustad Just fixed the build warning.  This is ready for review.

rraustad commented on src/EnergyPlus/HVACManager.cc:
```diff
@@ -1619,7 +1619,7 @@ namespace HVACManager {
 				if ( SimZoneEquipment ) {
 					if ( ( IterAir == 1 ) && ( ! FlowMaxAvailAlreadyReset ) ) { // don't do reset if already done in FirstHVACIteration
-						ResetTerminalUnitFlowLimits();
+						// ResetTerminalUnitFlowLimits(); // don't do reset at all - interferes with convergence and terminal unit flow controls
```
Agreed. MaxAvail reset first time through on FirstHVACIteration = true and be done with it.

rraustad commented on src/EnergyPlus/HVACManager.cc:
```diff
@@ -1812,6 +1812,7 @@ namespace HVACManager {
 								for ( ZonesCooledIndex = 1; ZonesCooledIndex <= AirToZoneNodeInfo( AirLoopIndex ).NumZonesCooled; ++ZonesCooledIndex ) {
 									TermInletNode = AirToZoneNodeInfo( AirLoopIndex ).TermUnitCoolInletNodes( ZonesCooledIndex );
 									Node( TermInletNode ).MassFlowRateMinAvail = Node( TermInletNode ).MassFlowRate * FlowRatio;
+									Node( TermInletNode ).MassFlowRateMaxAvail = max(Node(TermInletNode).MassFlowRateMaxAvail, Node(TermInletNode).MassFlowRateMinAvail);
```
I did see times when MinAvail > MaxAvail so this is needed.

rraustad commented on src/EnergyPlus/SingleDuct.cc:
```diff
@@ -2398,7 +2398,7 @@ namespace SingleDuct {
 		// if the massflow is below the minimum or greater than the Max it is set to either the Min
 		// or the Max as specified for the VAV model.
-		if ( ( QTotLoad < 0.0 ) && ( SysInlet( SysNum ).AirMassFlowRateMaxAvail > 0.0 ) && ( TempControlType( ZoneNum ) != SingleHeatingSetPoint ) ) {
+		if ( ( QTotLoad < 0.0 ) && ( SysInlet( SysNum ).AirMassFlowRateMaxAvail > 0.0 ) && ( TempControlType( ZoneNum ) != SingleHeatingSetPoint ) && ( GetCurrentScheduleValue( Sys( SysNum ).SchedPtr ) > 0.0 ) ) {
```
Added TU schedule to calculate air flow only when scheduled on.

rraustad commented on testfiles/5ZoneWLHPPlantLoopTower.idf:
```diff
@@ -1819,7 +1819,7 @@
     ,                        !- Watts per Zone Floor Area {W/m2}
     ,                        !- Watts per Person {W/person}
-    0.2,                     !- Return Air Fraction
+    ,                        !- Return Air Fraction
```
curious of the reason for this change. this would change RA temp and lead to additional iterations.

mjwitte replied:
Because these are zonehvac WLHPs which don't have any return air flow, so it was generating a warning about that and putting the heat gain into the zone anyway.

rraustad commented on testfiles/5ZoneWLHPPlantLoopTower.idf:
```diff
@@ -2083,52 +2035,13 @@
     ,                        !- Controller List Name
     AHU 1 Availability Managers,  !- Availability Manager List Name
-    autosize,                !- Design Supply Air Flow Rate {m3/s}
+    2.0,                     !- Design Supply Air Flow Rate {m3/s}
```
You changed this to fully hard sized. Isn't this the opposite of what example files should be?

mjwitte replied:
Yes, fully autosized would be nice.  Re-checking that, but I don't think it worked.  Yep, it hangs.  See #5544 (thanks for the reminder to post a new issue).

rraustad commented on tst/EnergyPlus/unit/SingleDuct.unit.cc:
```diff
@@ -0,0 +1,404 @@
+		//  4   Zone 1 Zone Equip Inlet
+
+		DataZoneEnergyDemands::ZoneSysEnergyDemand.allocate(1);
+		DataZoneEnergyDemands::ZoneSysEnergyDemand( 1 ).RemainingOutputRequired = -2000.0; // Zone 2 has a cooling load
```
// Zone 1 has a cooling load...but I don't really care about the innaccurate comment

mjwitte replied:
Actually, this line (and another like it) aren't needed anymore.

rraustad commented on tst/EnergyPlus/unit/SingleDuct.unit.cc:
```diff
@@ -0,0 +1,404 @@
+		SingleDuct::InitSys(SysNum, FirstHVACIteration); // Run thru init a second time with FirstHVACIteration set to false
+		SingleDuct::SimVAV( SysNum, FirstHVACIteration, ZoneNum, ZoneNodeNum );
+		EXPECT_EQ( 0.0, SingleDuct::SysOutlet( SysNum ).AirMassFlowRateMaxAvail );
+		EXPECT_EQ( 0.0, SingleDuct::SysOutlet( SysNum ).AirMassFlowRate );
```
So if the TU is scheduled off it has 0 flow. What if there is a minimum stop > 0? I guess this means the damper goes fully closed regardless of minimum stop setting. I'm OK with this, if you want flow at minimum stop, don't schedule off the TU.

rraustad commented on testfiles/5ZoneWaterLoopHeatPump.idf:
```diff
@@ -2909,7 +2612,7 @@
     0,                       !- Nominal Time for Condensate Removal to Begin {s}
     0,                       !- Ratio of Initial Moisture Evaporation Rate and Steady State Latent Capacity {dimensionless}
-    3000000,                 !- High Pressure Cutoff {Pa}
+    4000000,                 !- High Pressure Cutoff {Pa}
```
Ran into a high pressure cutoff didn't you. Been there done that.

mjwitte replied:
Yes.  Followed the lead of another similar example file (don't remember which one now).

mjwitte committed:
SingleDuct.unit.cc cleanup

rraustad commented:
The defect file shows the TU Zone Air Terminal Minimum Air Flow Fraction as a constant. I would think this would be the min of this or flow fraction. All other scheduling issues look correct now. Miami TMY3 but shouldn't matter.

![4825_failedminflowfrac](https://cloud.githubusercontent.com/assets/5778156/13703237/81dbee1e-e762-11e5-8f4c-1dffb640c31d.png)

rraustad commented:
Above based on TU schedule:

```
Schedule:Compact,
 school year,             !- Name
 Fraction,                !- Schedule Type Limits Name
 Through: 1/15,           !- Field 1
 For: AllDays,            !- Field 2
 Until: 24:00,0,          !- Field 3
 Through: 6/10,           !- Field 5
 For: AllDays,            !- Field 6
 Until: 24:00,1,          !- Field 7
 Through: 9/10,           !- Field 9
 For: AllDays,            !- Field 10
 Until: 24:00,0,          !- Field 11
 Through: 12/20,          !- Field 13
 For: AllDays,            !- Field 14
 Until: 24:00,1,          !- Field 15
 Through: 12/31,          !- Field 17
 For: AllDays,            !- Field 18
 Until: 24:00,0;          !- Field 19
```

mjwitte commented:
@rraustad I'm not following the comment.  What value looks wrong and when?

rraustad commented:
Maybe you didn't see the figures? View it on GitHub. Just below the top figure (VAV Damper Position) is the second figure with a dotted blue/orange line. This is the report variable "Zone Air Terminal Minimum Air Flow Fraction " before and after your fix. Why doesn't min flow frac go to 0 when the TU is scheduled off? If that report variable is static, it shouldn't be a report variable.

mjwitte commented:
i understand now.  That's a good question.  The min fraction can vary by schedule, so it's worthy of an output variable.  So, should it go to zero when the airterminal is off?  it's also showing a value when the entire air loop is off (made an alternate file with night cycle set to StayOff).  Is there a similar type of output variable we can look at for comparison?

rraustad commented:
Well, I guess this is OK. You can't maintain a minimum flow fraction if the TU is off. There is the:

HVAC,Average,Air System Outdoor Air Flow Fraction []
HVAC,Average,Air System Outdoor Air Minimum Flow Fraction []

1.29.6.2 Zone Air Terminal Minimum Air Flow Fraction []

This output is the value for the minimum air flow fraction setting on a VAV terminal. This is the average value over the frequency being reported. The minimum
air flow fraction is defined as the lower limit on the terminal unit air mass flow rate divided by the terminal units maximum air mass flow rate.

mjwitte commented:
Have you looked at an example file to see what HVAC,Average,Air System Outdoor Air Minimum Flow Fraction [] reports when an airloop is off?

rraustad commented:
It goes to 0 when flow is 0.

![5zoneaircooled_oaminflowfrac](https://cloud.githubusercontent.com/assets/5778156/13711916/f652e4a6-e78e-11e5-9f14-d2c9024fd479.png)

mjwitte commented:
So, that would argue that the terminal unit min fraction should also go to zero when flow is zero.  If you agree I'll change that and check to see if it's used in other terminal units.

rraustad commented:
I'd say I would not strongly disagree. I'm not sure what should happen to a schedule value used in a child component.

mjwitte committed:
Zero Zone Air Terminal Minimum Air Flow Fraction when no flow

mjwitte commented:
@rraustad Made the change to report zero for Zone Air Terminal Minimum Air Flow Fraction when there is no airflow. Note that the hourly output for the test file shows some odd numbers during unoccupied hours due to night cycle operation which is only on for a fraction of the hour.

mjwitte committed:
Fix padding in SingleDuct.hh

mjwitte committed:
Cleanup indents in SingleDuct.unit.cc

mjwitte committed:
More for Zone Air Terminal Minimum Air Flow Fraction

mjwitte commented:
Added the missing sets for ZoneMinAirFracReport.  Also added SetUpOutput for this variable for AirTerminal:SingleDuct:VAV:NoReheat which has similar inputs, but did not report this.

rraustad commented:
Defect file now shows minimum air flow fraction going to 0 when air loop is off.

![defect_4285](https://cloud.githubusercontent.com/assets/5778156/13753635/1cb446dc-e9ea-11e5-978f-b42d20e79796.png)

rraustad commented:
The diff's here are the same as those for https://github.com/NREL/EnergyPlus/pull/5532 and those have been reviewed. Considering this issue closed.

rraustad commented:
Always good to merge before closing.
